## Leonardo Rodriguez Hernandez ##  
## 04 - Marzo - 2024 ##

# Examen A Módulo 1 - Optimización.

**Lea cuidadosamente las siguientes indicaciones antes de comenzar el examen:**
- Para resolver el examen edite este mismo archivo y renómbrelo de la siguiente manera: *Examen1_ApellidoNombre*, donde *ApellidoNombre* corresponde a su apellido paterno con la inicial en mayúscula, seguido de su primer nombre con la inicial en mayúscula **sin acentos**. Por ejemplo, en mi caso el nombre del archivo sería *Examen1_MoralesGabriel*.
- Resuelva los puntos en el espacio provisto. Si requiere agregar más celdas para código o escritura, hágalo (`Esc`,`B`) o (`Esc`,`A`).
- Recuerde que también se está evaluando su capacidad de interpretar los resultados. Escriba sus interpretaciones/conclusiones en celdas utilizando *Markdown*.
- El formato de presentación de todo el examen en general debe ser adecuado. Use tamaños de letra, colores, etiquetas, etcétera.
- No se resuelven dudas de ningún tipo por los profesores de la asignatura. Por favor, absténgase de preguntar.
- Revise bien el tipo de examen. Sólo se le calificará si realiza el tipo de examen asignado.

---
# Primer punto. Optimización de funciones con `SymPy` (37 puntos)

**Costo mínimo**. Imagine dos contenedores cilíndricos sin tapa que deben satisfacer las siguientes condiciones:

- El primer contenedor tiene el doble de altura que el segundo.
- La altura del primer contenedor (el de mayor tamaño) es el doble de su radio.
- El material para la base cuesta 10.00 MXN por cada $m^2$.
- El material para los lados cuesta 6.00 MXN por cada $m^2$.

¿Cuál es el costo del contenedor más barato?


1. Escriba el volumen $V_1(r_1)$ y $V_2(r_2, r_1)$ de ambos contenedores como dos funciones distintas, donde $r_1$ es el radio del primer contenedor y $r_2$ es el radio del segundo. Realice este punto usando $\LaTeX$ y celdas *Markdown*.


Si el volumen de un cilindro es:  
$ V=πr^2h$  
### Volumen del primer contenedor $V_1(r_1)$:

$ V_1(r_1) = \pi r_1^2 h_1 $

Sustituyendo $h_1 = 2r_1$, obtenemos:

$ V_1(r_1) = \pi r_1^2 (2r_1) $

### Volumen del segundo contenedor $V_2(r_2, r_1)$:


 $V_2(r_2, r_1) = \pi r_2^2 (2r_1)$


2. Escriba el costo **C** de fabricar ambos contenedores como una función de los radios de las bases. Realice este punto usando $\LaTeX$ y celdas *Markdown*.

$C(r_1,r_2) = 10πr_1^2 + 12πr_1(2r_1) + 10πr_2^2 + 12πr_2(2r_2)$  

3. Determina las dimensiones de los contenedores si el volumen del contenedor pequeño debe ser de $12m^3$ y el costo debe ser mínimo.

In [46]:
import sympy as sp
import nu
r1 = sp.var("r1")
r2 = sp.var("r2")
h1 = 2 * r1
h2 = 0.5 * h1
V1 = sp.pi * r1**2 * h1
V2 = sp.pi * r2**2 * h2
pi = np.pi
C = 10 * (sp.pi * r1**2 + sp.pi * r2**2) + 12 * (sp.pi * r1 * h1 + sp.pi * r2 * h2)

restriccion = Eq(V2, 12)

r2enr1 = solve(restriccion, r2)[1]

C_r1 = C.subs(r2, r2enr1)

dC_r1 = diff(C_r1, r1)

r1_critico = solve(dC_r1, r1)[0]

r2_critico = r2enr1.subs(r1, r1_critico)

print("Radio del primer contenedor: ",r1_critico)
print("Radio del segundo contenedor: ",r2_critico)

Radio del primer contenedor:  0.764313731117990
Radio del segundo contenedor:  2.23552667172687


3. ¿Cuál fue el costo de fabricar este par de contenedores?

In [47]:
Costo = C.subs({r1: r1_critico, r2: r2_critico})
print("El costo minimo es de: ", Costo)


El costo minimo es de:  90.3414860833656*pi


---
# Segundo punto. Programación lineal (37 puntos)

Una empresa se dedica a fabricar tarjetas de control para automóviles, y están en su momento de planeación de compras anuales. Hay 5 proveedores que pueden proveer los componentes, cada uno a su propio precio.

<table>
  <tr>
      <th>  </th>
      <th>Proveedor 1</th>
      <th>Proveedor 2</th>
      <th>Proveedor 3</th>
      <th>Demanda</th>
  </tr>
  <tr>
      <th>Componente 1</th>
      <td>0.5</td>
      <td>0.6</td>
      <td>0.7</td>
      <td>15000</td>
  </tr>
  <tr style="border-bottom: 1px solid #000;">
      <th>Componente 2</th>
      <td>0.7</td>
      <td>0.5</td>
      <td>0.6</td>
      <td>20000</td>
  </tr>
  <tr style="border-bottom: 1px solid #000;">
      <th>Componente 3</th>
      <td>0.32</td>
      <td>0.54</td>
      <td>0.8</td>
      <td>12000</td>
  </tr>

  <tr style="border-bottom: 1px solid #000;">
      <th>Capacidad</th>
      <td> 21000 </td>
      <td> 16000 </td>
      <td> 16000 </td>
      <td>  </td>
  </tr>
</table>

El proveedor tiene restricciones de capacidad de fabricación del tipo de componentes, por lo que un proveedor no puede satisfacer nuestra demanda completa para cada componente.

Se desea conocer la cantidad que se debe comprar de cada componente a cada proveedor para gastar la menor cantidad de dinero posible, pero al mismo tiempo lograr satisfacer la demanda para poder continuar con la manufactura del producto.

1. Escriba la función a **minimizar** junto con las restricciones, explicando detalladamente cada paso (usando fórmulas en $\LaTeX$, en celdas Markdown). Además escriba el problema en la forma \begin{equation}
\begin{array}{ll}
\min_{\boldsymbol{x}} & \boldsymbol{c}^T\boldsymbol{x} \\
\text{s. a. }        & \boldsymbol{A}_{eq}\boldsymbol{x}=\boldsymbol{b}_{eq} \\
                     & \boldsymbol{A}\boldsymbol{x}\leq\boldsymbol{b},
\end{array}\qquad\text{(23 puntos)}
\end{equation}

**Criterio de optimizacion**  
Minimizar el gasto.

**Variables**  
$P_1 = Componente 1 de Proovedor 1$  
$P_2 = Componente 1 de Proovedor 2$  
$P_3 = Componente 1 de Proovedor 3$  

$P_4 = Componente 2 de Proovedor 1$  
$P_5 = Componente 2 de Proovedor 2$  
$P_6 = Componente 2 de Proovedor 3$   

$P_7 = Componente 3 de Proovedor 1$  
$P_8 = Componente 3 de Proovedor 2$  
$P_9 = Componente 3 de Proovedor 3$   

**Restricciones de demanda**    
$P_1 + P_2 + P_3 = 15000 $  
$P_4 + P_5 + P_6 = 20000 $  
$P_7 + P_8 + P_9 = 12000 $  

**Restricciones de capacidad**  
$P_1 + P_4 + P_7 \leq 21000 $  
$P_2 + P_5 + P_8 \leq 16000 $  
$P_3 + P_6 + P_9 \leq 16000 $  

**Funcion objetivo**  
$0.5P_1 + 0.6P_2 + 0.7P_3 + 0.7P_4 + 0.5P_5 + 0.6P_6 + 0.32P_7 + 0.54P_8 + 0.8P_9$

In [3]:
import numpy as np
c= np.array([0.5,0.6,0.7,0.7,0.5,0.6,0.32,0.54,0.8])
Aub = np.array([
               [-1,-1,-1,0,0,0,0,0,0],
               [0,0,0,-1,-1,-1,0,0,0],
               [0,0,0,0,0,0,-1,-1,-1],
])
bub = np.array([[-15000,-20000,-12000]
])

Aeq = np.array ([
    [1,0,0,1,0,0,1,0,0],
    [0,1,0,0,1,0,0,1,0],
    [0,0,1,0,0,1,0,0,1]
])
Beq = np.array ([
    [21000,16000,16000]
])
import scipy.optimize as opt
opt.linprog(c, Aub, bub,Aeq,Beq)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 26060.0
              x: [ 3.000e+03  0.000e+00  1.200e+04  0.000e+00  1.600e+04
                   4.000e+03  1.800e+04  0.000e+00  0.000e+00]
            nit: 6
          lower:  residual: [ 3.000e+03  0.000e+00  1.200e+04  0.000e+00
                              1.600e+04  4.000e+03  1.800e+04  0.000e+00
                              0.000e+00]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  3.000e-01
                              0.000e+00  0.000e+00  0.000e+00  1.200e-01
                              2.800e-01]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0

2. Resuelva el problema utilizando la función `linprog`. ¿Cómo debe ser la distribución de compras? ¿Cuánto dinero se necesita para satisfacer la demanda? (10 puntos)

**Los resultados dicen que:**    
    - Del componente 1 se necesitan de 3 mil del primer proveedor y 12 mil del 3er proveedor.  
    - Del componente 2 se necesitan de 16 mil del 2do proveedor y 4 mil del 3er proveedor.  
    - Del componente 3 se necesitan de 18 mil del 1er proveedor.  

El costo total es de: **26060** u.m.  

# Tercer punto. Ajuste de curvas (36 puntos)

Utiliza el archivo adjunto en Canvas en el apartado que corresponde al examen.

Escribe tu número de expediente: 

Lee el archivo y extrae la siguiente información:
- $x$ es la primer columna del archivo
- $y$ es la columna correspondiente a los últimos dos dígitos de tu número de expediente + 1. E.g. mi expediente es $693871$; por lo tanto, debo usar la columna $72$.

1. Ajuste los datos con polinomios de grados 1, 3, y 5. Muestre los errores de forma gráfica. ¿Qué polinomio se ajusta mejor?

2. Muestre en un gráfico los datos y el polinomio que mejor se ajuste.

3. Use el modelo para averiguar el valor de $y$ cuando:
- $x=0.001$
- $x=-0.1$
- $x=2.1$

4. ¿Qué diferencias encuentra en cuanto a la interpolación y la extrapolación?

5. Agregue un factor $\cos{(10x)}$ a su modelo y grafique los resultados. ¿Mejoró el resultado?